File: 793. Preimage Size of Factorial Zeroes Function/main.py
import sys


class Solution:
    def preimageSizeFZF(self, k: int) -> int:
        return right(k) - left(k) + 1


def left(target: int) -> int:
    left: int = 0
    right: int = pow(10, 9)
    while left < right:
        mid: int = int(left + (right - left)/2)
        z = zeros(mid)
        if z < target:
            left = mid+1
        elif z > target:
            right = mid
        else:
            right = mid

    return left


def right(target: int) -> int:
    left: int = 0
    right: int = pow(10, 9)
    while left < right:
        mid: int = int(left + (right - left)/2)
        z = zeros(mid)
        if z < target:
            left = mid+1
        elif z > target:
            right = mid
        else:
            left = mid+1
    return left-1


def zeros(n: int) -> int:
    divisor = 5
    res: int = 0
    while divisor <= n:
        res += n / divisor
        divisor *= 5
    return int(res)

File: 793. Preimage Size of Factorial Zeroes Function/main.go
package main

import "math"

func main() {
	preimageSizeFZF(1000000000)
}

func preimageSizeFZF(k int) int {
	return int(right(k)-left(k)) + 1
}

func left(target int) int64 {
	var l, r int64 = 0, math.MaxInt64
	t := int64(target)

	for l < r {
		mid := l + (r-l)/2
		z := zeros(mid)

		if z < t {
			l = mid + 1
		} else if z > t {
			r = mid
		} else {
			r = mid
		}
	}
	return l
}

func right(target int) int64 {
	var l, r int64 = 0, math.MaxInt64
	t := int64(target)

	for l < r {
		mid := l + (r-l)/2
		z := zeros(mid)

		if z < t {
			l = mid + 1
		} else if z > t {
			r = mid
		} else {
			l = mid + 1
		}
	}
	return l - 1
}

func zeros(n int64) int64 {
	var res int64 = 0
	for d := n; d/5 > 0; d /= 5 {
		res += d / 5
	}
	return res
}

//1000000000
